great_expectations_provider/operators/great_expectations.py

```diff
@@ -267,32 +280,70 @@ def make_connection_string(self) -> str:
 
             uri_string = f"snowflake://{self.conn.login}:{self.conn.password}@{snowflake_account}.{snowflake_region}/{snowflake_database}/{self.schema}?warehouse={snowflake_warehouse}&role={snowflake_role}"  # noqa
 
-            # private_key_file is optional, see:
-            # https://docs.snowflake.com/en/user-guide/sqlalchemy.html#key-pair-authentication-support
-            snowflake_private_key_file = self.conn.extra_dejson.get(
-                "private_key_file", self.conn.extra_dejson.get("extra__snowflake__private_key_file")
-            )
-            if snowflake_private_key_file:
-                uri_string += f"&private_key_file={snowflake_private_key_file}"
         elif conn_type == "gcpbigquery":
             uri_string = f"{self.conn.host}{self.schema}"
         elif conn_type == "sqlite":
             uri_string = f"sqlite:///{self.conn.host}"
         # TODO: Add Athena and Trino support if possible
         else:
             raise ValueError(f"Conn type: {conn_type} is not supported.")
-        return uri_string
+        return {"connection_string": uri_string}
+
+    def build_snowflake_connection_config_from_hook(self) -> Dict[str, str]:
+        from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+        from cryptography.hazmat.backends import default_backend
+        from cryptography.hazmat.primitives import serialization
+
+        hook = SnowflakeHook(snowflake_conn_id=self.conn_id)
+
+        # Support the operator overriding the schema
+        # which is necessary for temp tables.
+        hook.schema = self.schema or hook.schema
+
+        conn = hook.get_connection(self.conn_id)
+        engine = hook.get_sqlalchemy_engine()
+
+        url = engine.url.render_as_string(hide_password=False)
+
+        private_key_file = conn.extra_dejson.get("extra__snowflake__private_key_file") or conn.extra_dejson.get(
+            "private_key_file"
+        )
+
+        if private_key_file:
+            private_key_pem = Path(private_key_file).read_bytes()
+
+            passphrase = None
+            if conn.password:
+                passphrase = conn.password.strip().encode()
+
+            p_key = serialization.load_pem_private_key(private_key_pem, password=passphrase, backend=default_backend())
+
+            pkb = p_key.private_bytes(
+                encoding=serialization.Encoding.DER,
+                format=serialization.PrivateFormat.PKCS8,
+                encryption_algorithm=serialization.NoEncryption(),
+            )
+            return {
+                # Unfortunately GE uses deepcopy when instantiating the SqlAlchemyExecutionEngine
+                # which uses pickle and SAEngine is not pickleable.
+                # "engine": engine,
+                "url": url,
+                "connect_args": {
+                    "private_key": pkb,
+                },
+            }
+
+        return {"url": url}
```

Comment:
I don't understand why this is "url" and not "connection_string" like the other connections. Can you point to the GX doc where this is the case?

Reply:
This is undocumented by GX afaik, for some reason it wasn't possible to supply connection args with a `connection_string`. See: https://github.com/great-expectations/great_expectations/blob/main/great_expectations/datasource/sqlalchemy_datasource.py#L256

On this line I could return `{"connection_string": url}` I wanted to keep it as consistent as possible with the key based code path but happy to change it.

Reply:
Ok, that looks good. Didn't see that bit that it checks for `connection_string` or `url` before.